sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -287,6 +288,17 @@ object ColumnPruning extends Rule[LogicalPlan] {
 }
 
 /**
+ * Combines the adjacent [[Project]] and [[Join]] operators, iff their output sets are the same.
+ */
+object CombineProjectJoin extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case p @ Project(_, j: Join)
+      if j.outputSet == p.outputSet => j
```

Comment:
hi @gatorsmile , as you said, this will lose the ordering and alias information, are you going to improve join operator to make it smarter on this?